### ISSUE TITLE
Question about the prune rate

### ISSUE THREAD

**Issue by RobertFlame** [allenai/hidden-networks#9]
Hi,

Recently I tried the code and found that there seems to have inconsistency of the definition of prune rate throughout the code.

In builder.py, when the scale_fan is True, the fan is scaled by a factor of (1-prune_rate), where it denotes the top k% of weights as indicated in the paper. So here prune_rate means the ratio of weights to be removed. This meaning is consistent with main.py line 318, which calculates the number of parameters in the subnetwork. 

But in conv_type.py lines 20-25, when generating the subnet mask, (1-prune_rate) of the weights are removed:
`j = int((1 - k) * scores.numel())
flat_out = out.flatten()
flat_out[idx[:j]] = 0
flat_out[idx[j:]] = 1`
 which seems incorrect. So I think this is a bug and line 20 should be changed to 
`j = int(k * scores.numel())`

Could you please clarify a bit what exactly does the prune_rate means in your code?

Thanks

**Comment by Lasitha-93** [comment]
Hi,

I'm also interested in this paper and trying to reproduce the results. I'm totally agree with your question because the relationship of prune_rate-->k is now very clear both in the paper and the code (as far as I understood) but I agreeing with your explanation of the relationship, what I think we need to focus is in conv_type.py-->GetSubnet();

        **# Get the subnetwork by sorting the scores and using the top k%
        out = scores.clone()
        _, idx = scores.flatten().sort()
        j = int((1 - k) * scores.numel())

        # flat_out and out access the same memory.
        flat_out = out.flatten()
        flat_out[idx[:j]] = 0
        flat_out[idx[j:]] = 1

        return out**

j = int((1 - k) * scores.numel()) where 'j' is used to slice the flat_out scores. But since it's sorted(Asc. by default), flat_out[idx[:j]] = 0 refers to the pruned scores and flat_out[idx[j:]] = 1 is giving the 'k' number of scores. Therefore, I think the code is correct.

Regards!

**Comment by RobertFlame** [comment]
@Lasitha-93 I agree that the logic of the code is correct. The problem is that the definition of prune rate (or k) in GetSubnet() and weight initialization (implemented in utils/builder.py in function _init_conv() ) is not consistent. So either the _init_conv() is wrong or GetSubnet() is wrong. However, I don't think it will affect the conclusion of the paper too much, because the authors mostly conduct the qualititive analysis on the experiments, not quantitative. In my opinion, after correcting the qualititive analysis of the paper will still be true. But it's always good to figure out the problem and fix it.

Please correct me if I misunderstand something ;)
